toptal/picasso
# fix: adjust for SSR

[FX-4048]

### Description

1. reverting the Root component render guard to not render null on the first render in the browser
2. `SkeletonLoader` - add `uniqueKey` property to disable generating random uid
3. `Page.TopBar` - conditional rendering of components can't be done by JS to match HTML output on server and browser

### How to test

to reproduce the issue in Client-Signup
1. clone the project locally `git clone https://github.com/toptal/client-signup`
2. `cd apps/admins`
3. `yarn install && yarn start:dev`
4. open `http://localhost:4201/` in the browser
5. open the console and you will see multiple errors/warnings about HTML mismatch 
---
6. kill the process
7. checkout to the testing branch `git checkout CC-868-picasso-update`
8. `yarn install && yarn start:dev`
9. open `http://localhost:4201/` in the browser
10. no errors should occur (only graphql and right now also https://github.com/toptal/picasso/pull/3612)

---
or you can test it in your own local NextJS application
- when creating a new app make sure to select the old format with `pages` so we can use MUI SSR example https://github.com/mui/material-ui/blob/v4.x/examples/nextjs/pages/_document.js (just copy _document and _app) add Picasso provider to the page and you will see it fail.

### Development checks


**Breaking change**

- n/a codemod is created and showcased in the changeset
- n/a test alpha package of Picasso in StaffPortal

> All **_development checks_** should be done and set checked to pass the
> **GitHub Bot: TODOLess** action

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run package:alpha-release` - Release alpha version
- `@toptal-anvil ping reviewers` - Ping FX team for review

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and

1. You have no extra requests.
12. You have optional requests.
   1. Add `nit:` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because

1. Something is broken after the changes.
4. Acceptance criteria is not reached.
13. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:

1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
5. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>


[FX-4048]: https://toptal-core.atlassian.net/browse/FX-4048?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
